### PR TITLE
[hal] Merge `Limits` and `Capabilities` as `PhysicalDeviceProperties`

### PR DESCRIPTION
**Pull request by tangmi**
This PR:
- Merges `hal::Limits` and `hal::Capabilities` as `PhysicalDeviceProperties`.
  - The name was chosen to be more similar to Vulkan.
- Adds properties structs for mesh shaders and descriptor indexing.
- Adds a `DescriptorLimits` struct to hold descriptor limits and makes the values `u32`.

PR checklist:
- [x] `make` succeeds (on *nix)
- [x] `make reftests` succeeds
- [x] tested examples with the following backends:
  - [x] bench: dx12, vulkan
  - [x] colour-uniform: dx12, vulkan
  - [x] compute: dx12, vulkan
  - [x] quad: dx12, vulkan

### CODE REVIEW

File: src/hal/src/lib.rs
Comment (tangmi) [inline src/hal/src/lib.rs:496]: This is a direct mapping from `vkPhysicalDeviceDescriptorIndexingPropertiesEXT`, which is probably more than what we want for `gfx`.

File: src/hal/src/lib.rs
Comment (kvark) [inline src/hal/src/lib.rs:530]: we have quite a few places for `max_XXX_descriptor_XXX_some_stage`. I wonder if we can make this nicer? I.e.
```rust
struct BindLimits {
  max_uniform_buffers: u32,
  max_samplers: u32,
  ...
}
pub per_stage_descriptor_update_limits: BindLimits,
pub descriptor set_update_limits: BindLimits,
```
We'd probably use it in the `Limits` too!

**Review by kvark — APPROVED** [review]
Very nice!

File: src/hal/src/lib.rs
Comment (tangmi) [reply src/hal/src/lib.rs:530]: I think we'll need to get rid of the `*_update_after_bind_*` entries here, since `gfx` doesn't (yet?) support that.

I also noticed most of the core `Limits` are `usize` when the actual Vulkan value is `u32`, is this intentional? And should I follow that pattern for mesh shader and descriptor indexing? The best reason I can come up with is that these values will likely be used in calculations to size rust vecs or arrays, which would operate on `usize`.

Comment (kvark) [reply src/hal/src/lib.rs:530]: Not intentional, we should probably change them to `u32`.

Comment (kvark) [reply src/hal/src/lib.rs:530]: So are we removing the `_update_after_bind_` for now?

**Review by kvark — APPROVED** [review]
LGTM, thank you for doing this!
Just one small question

File: src/hal/src/lib.rs
Comment (tangmi) [reply src/hal/src/lib.rs:530]: Admittedly I'm not sure what we'd need for this feature. I've removed the update-after-bind stuff because gfx itself doesn't support that feature. I believe update-after-bind is pretty important for descriptor indexing--although, again, I could be wrong here.

@cwfitzgerald I recall we needed descriptor indexing to handle a bug in Intel drivers for WebGPU... can I defer to you for what is important to expose as a part of `hal` for descriptor indexing? Were you imagining this being a mostly "private capabilities"-type feature or something publicly exposed?

Comment (kvark) [reply src/hal/src/lib.rs:530]: this is the stuff from here:
> VK_DESCRIPTOR_BINDING_UPDATE_AFTER_BIND_BIT indicates that if descriptors in this binding are updated between when the descriptor set is bound in a command buffer and when that command buffer is submitted to a queue, then the submission will use the most recently set descriptors for this binding and the updates do not invalidate the command buffer. Descriptor bindings created with this flag are also partially exempt from the external synchronization requirement in vkUpdateDescriptorSetWithTemplateKHR and vkUpdateDescriptorSets. Multiple descriptors with this flag set can be updated concurrently in different threads, though the same descriptor must not be updated concurrently by two threads. Descriptors with this flag set can be updated concurrently with the set being bound to a command buffer in another thread, but not concurrently with the set being reset or freed.

We don't have that supported yet, so we can leave this off PR.
However, it would be nice to at least group those descriptor limits together in a struct, so that if/when we need to add another set of limits for the "update_after_bind", we can just re-use the same struct type.

Comment (tangmi) [reply src/hal/src/lib.rs:530]: I'll do another pass to see what I can group together... I remember the last time I looked at it (briefly!) every group of limits were just slightly different from another and didn't meaningfully have an abstraction to pull out.

I'd also like to wait on @cwfitzgerald just to make sure this is sufficient for the Intel/WebGPU work?

Comment (tangmi) [reply src/hal/src/lib.rs]: Also removed the _update_after_bind_ fields

**Review by kvark — APPROVED** [review]
Could we squash some of the commits (that don't build independently) before merging?

File: src/hal/src/lib.rs
Comment (cwfitzgerald) [reply src/hal/src/lib.rs]: So for update after bind, the idea is that it actually introduces no more requirements on the API to just use update after bind descriptors like normal descriptors. i.e. update after bind descriptors are a superset of non-update after bind descriptors, so we should just implicitly use them to raise the limits on intel.

**Comment by kvark** [comment]
@tangmi Let's rebase and merge

**Review by kvark — APPROVED** [review]
We'll need to add a changelog entry at some point, but everything looks good!
bors r+

**Pull request merged**
